Feature/ignore unreachable ice server

sjkummer opened the pull request:
Adding an additional nat config item 'ignore_unreachable_ice_server' which allows to start Janus on edge devices even if WAN connectivity is temporarily unavailable. The default behaviour (without the new config item) remains the same as before.
https://github.com/meetecho/janus-gateway/issues/987

lminiero commented:
This has been discussed in the past, and I'm not sure it's wise. That startup test can be very helpful to figure out if the connection details will be usable, and allowing people to skip it will cause issues in some cases: what if it's not just "temporarily unreachable", but unreachable-period, and you're starting the server anyway? Besides, in case of STUN it means Janus won't have access to the public IP, which shouldn't be an issue in most cases (it's only used for nat-1-1, which should not be used in conjunction with STUN anyway), but still implies incomplete information when you ask Janus for info.

Anyway, if you guys really want to shoot yourself in the foot, I'll think about it. That said, please sign the [CLA](https://cla.conf.meetecho.com/janus/), as it's required for merging.

sjkummer commented:
Hello Lorenzo, thanks for your fast response. 

I fully agree that reachability problems should not be ignored if Janus runs as cloud server. I have added a (hopefully) very clear warning in the example-config and added a note about using the systemd 'After' directive as best-practice. Further, I have signed the CLA.

I really think this change should be merged to make Janus ready for Edge-Computing and IoT use cases. As an example: Janus can be used to add WebRTC-capabilities to legacy audio/video streams in order to make them interoperable with modern browsers. This streams can be consumed by on-site applications (which do not require TURN) and optionally remote clients (which require TURN). Why should we disallow the on-site usage of an application just because the internet connectivity (of the company / factory / organisation / public office / ... ) is unreliable?

This change should improve the reliability and the range of application for Janus and its community.

Best regards,
Sebastian

lminiero commented:
> Why should we disallow the on-site usage of an application just because the internet connectivity (of the company / factory / organisation / public office / ... ) is unreliable?

We wouldn't be preventing anything: if the application doesn't require TURN, they wouldn't put any TURN server in the config, and there would be no attempt to connect to anything, and so no risk of getting a failure doing that, whether they're connected to the internet or not. This can only happen if one actively sets a STUN or TURN server in the settings.

Notice that skipping the STUN/TURN check because it fails does NOT mean they won't be used: they'll still be involved any time a PeerConnection needs to be established, because they were in the configuration. So, just for the sake of clarity, if the purpose of this PR is disabling STUN if that check fails and just go on, that's not what would happen: the STUN/TURN server would still be in the configuration properties, and would be used accordingly (and possibly fail again). If that's not the purpose, then that's ok, and this can be merged.

lminiero commented:
PS: to clarify, the STUN server would still be attempted for a new PeerConnection if what failed was the STUN connectivity check; if the failure was in resolving the address, then Janus would not have reached the part of the code where it takes note of the STUN server address, and so it would NOT be used further. For TURN, we never to any connectivity check, and so the only failure can come from an error trying to resolve the address (which means that TURN would indeed be always be disabled, in case of a failure at startup).

lminiero commented:
To solve this ambiguity, I think the stun/turn init methods, before returning the error, should free the strings containing the resolved address. This way, no matter the cause, if the server is configured to go on the user knows STUN/TURN support are disabled because of that.

sjkummer commented:
> PS: to clarify, the STUN server would still be attempted for a new PeerConnection if what failed was the STUN connectivity check; if the failure was in resolving the address, then Janus would not have reached the part of the code where it takes note of the STUN server address, and so it would NOT be used further. For TURN, we never to any connectivity check, and so the only failure can come from an error trying to resolve the address (which means that TURN would indeed be always be disabled, in case of a failure at startup).

Thanks for your further clarifications! My original intention was to use the configured STUN and TURN servers for all new PeerConnections after the ice servers have become reachable. But I realise that this actually would mean to resolve the ice server addresses at a later point (again) which would increase the complexity of this change. 
I agree to free up the resolved STUN address after the connectivity check failed to achieve a consistent behaviour. I'll push a change.

lminiero commented:
Thanks! It shouldn't just be set to NULL, though: the string is strdup-ed and do should be freed. The same should be done in the TURN setup too, since the new ignore property is involved for that method was well.

As a side note, it looks like there are some conflicts now?

sjkummer commented:
Okay, I have resolved the conflicts and freed the stun server string.
For TURN, no string needs to be freed, because no further test is done after the TURN server address was resolved, right?

lminiero commented on ice.c:
```diff
@@ -1010,10 +1010,19 @@ int janus_ice_set_stun_server(gchar *stun_server, uint16_t stun_port) {
 	janus_network_address public_addr = { 0 };
-	if(janus_ice_test_stun_server(&addr, janus_stun_port, 0, &public_addr, NULL) < 0)
+	if(janus_ice_test_stun_server(&addr, janus_stun_port, 0, &public_addr, NULL) < 0) {
+		if(janus_stun_server) {
```
No need for the `if`, `g_free` does nothing if it's `NULL`.

sjkummer replied:
good to know :-)

lminiero commented on ice.c:
```diff
@@ -1010,10 +1010,19 @@ int janus_ice_set_stun_server(gchar *stun_server, uint16_t stun_port) {
+	}
 	if(janus_network_address_to_string_buffer(&public_addr, &addr_buf) != 0) {
 		JANUS_LOG(LOG_ERR, "Could not resolve public address...\n");
+		if(janus_stun_server) {
```
Same as above.

lminiero commented on janus.c:
```diff
@@ -4235,12 +4238,20 @@ gint main(int argc, char *argv[])
+			JANUS_LOG(LOG_FATAL, "Invalid STUN address %s:%u\n", stun_server, stun_port);
+			exit(1);
+		} else {
+			JANUS_LOG(LOG_ERR, "Invalid STUN address %s:%u\n", stun_server, stun_port);
```
I'd add a "STUN will be disabled" text to the error, so that it's more explicit.

lminiero commented on janus.c:
```diff
@@ -4235,12 +4238,20 @@ gint main(int argc, char *argv[])
+			JANUS_LOG(LOG_FATAL, "Invalid TURN address %s:%u\n", turn_server, turn_port);
+			exit(1);
+		} else {
+			JANUS_LOG(LOG_ERR, "Invalid TURN address %s:%u\n", turn_server, turn_port);
```
Same here.

lminiero commented:
> For TURN, no string needs to be freed, because no further test is done after the TURN server address was resolved, right?

You're right! I should probably add a STUN check for TURN servers too, to make sure they're reachable (as a TURN server is a STUN server too), but I can do that after this is merged.

sjkummer commented:
Should all be cleaned up now.

lminiero commented:
Thanks for your patience and quick fixes, looks good now! Merging then :+1:

sjkummer commented:
Great thanks 🙂
